### PR TITLE
Feature/home page react (#1)

### PR DESCRIPTION
**Pull request by jenibelle08**
## Description

This PR converts the home page to React. The conversion keeps the functionality and appearance of the home page unchanged, but contains substantial changes to how these pages are composed. This is an interim step to implementing the redesign of the home page.

*This PR is not intended for merge into master until the completion of the homepage revamp*

## Notes

In addition to the conversion of the content of home page, this PR also converts the header and footer of the site to React, also in anticipation of the home page redesign. 

New React apps were added in the homepage (rendered by `homepage/main.ts`) and the base template (rendered in `base/main.ts`). 

To facilitate different modules independently rendering components on load, the `window.onLoad` assignments have in all modules been converted to `window.addEventListener` (without this change, only the last assigned onLoad handler is applied).

We set up a system of standardized transferring of routing and translation label information from the Jinja templates to the TypeScript. The routing and translations are now rendered into respective elements in an auxiliary template that is included by the base template. These can then be read by any TypeScript module (such as a `main.ts`) and then passed through to to the relevant apps. Individual pages can still independently read their own attributes past up through the templates.

## Considerations

Previously, the menu content in the headers and footers were contained directly in the templates. The move to rendering with React components presents an opportunity separate this content from the actual implementation of the header and footer, with the goal of making it easier for forks to override the content without having to make any changes to the component that implements it.

We have tentatively set up a process by which the header and footer content are sent through from the backend as JSON (in a similar way to how the partners and topics are sent through to the homepage). 

The goal here will be to set up an easy way for a fork to provide a custom version of the header or footer in their own backend (for example, with the commonly used `_custom_dc` tag in the filename), to override in the fork without affecting the core. This could be done either in a separate PR, or following feedback and discussion, in this PR.

### Future Fork Customization

Before this project is completed, we will need to consider the potential fork customization options for the new React structure of the home page (as we are no longer dealing with a scenario where a significant amount of the material to be customized lies in the template).

We can consider three levels of customization that a fork might want to implement:
* Content: This is the simplest customization, referred to above, where they may which to change the available menus and associated text.
* Style: In this scenario, a fork may which to keep the structure of the components intact, but style the components (changing fonts, colors, and the layout to the degree that the structure allows).
* Components: Here, a fork may choose to remove, edit or otherwise substantially change the components themselves.

These will be considered as development of the homepage revamping progresses.

### CODE REVIEW

File: static/js/apps/base/main.ts
Comment (dwnoble) [inline static/js/apps/base/main.ts:45]: This might need to get lowercased-- im seeing the full footer in other pages in the app: 

in http://localhost:8080/tools/statvar

```
<div id="metadata-base" class="d-none"
...
  data-hide-full-footer="True"
...
/>
```

File: server/templates/base.html
Comment (dwnoble) [inline server/templates/base.html:192]: It's important to keep all of the `{#  TRANSLATORS: ... #}` notes when migrating the `{% trans %}...{% endtrans %}` tags

These tags get extracted by pybabel tool and wind up in the translation files: https://github.com/datacommonsorg/website/blob/f5e8e87c2291d87dfa37a3a887f01d7ff28d6467/server/i18n/all.pot#L77

References:
https://github.com/datacommonsorg/website/blob/master/docs/localization.md
https://github.com/datacommonsorg/website/blob/f5e8e87c2291d87dfa37a3a887f01d7ff28d6467/scripts/extract_messages.sh#L37

File: server/templates/auxiliary/labels.html
Comment (dwnoble) [inline server/templates/auxiliary/labels.html:18]: Make sure to include `{# TRANSLATORS #}` comments (see my other comment about this)

File: server/templates/base.html
Comment (dwnoble) [inline server/templates/base.html:107]: Nit: maybe delete this flag in a follow-on pr - i dont think it's actually used anymore

File: static/js/apps/base/utilities/utilities.ts
Comment (dwnoble) [inline static/js/apps/base/utilities/utilities.ts:19]: Nit: add a comment describing this function

File: static/js/apps/base/utilities/utilities.ts
Comment (dwnoble) [inline static/js/apps/base/utilities/utilities.ts:32]: Nit: add a comment describing this function

File: static/js/apps/base/utilities/utilities.ts
Comment (dwnoble) [inline static/js/apps/base/utilities/utilities.ts:69]: A typo when fetching a label might be hard to detect. Could you add a log message if the label isn't found?

File: static/js/apps/base/utilities/utilities.ts
Comment (dwnoble) [inline static/js/apps/base/utilities/utilities.ts:65]: Should this be `Proxy<Labels>`?

File: static/js/apps/base/utilities/utilities.ts
Comment (dwnoble) [inline static/js/apps/base/utilities/utilities.ts:48]: i think this should throw an error if the route isn't found - does that sound right?

**Review by dwnoble — COMMENTED** [review]
Thanks Jennifer! The behavior looks overall consistent with the existing homepage. I did see the other pages are loading the full footer though:
Screenshot 2024-09-03 at 4 57 29 PM

should be:

Screenshot 2024-09-03 at 4 57 29 PM


I'll let @beets  weigh in as well

File: static/js/apps/base/utilities/utilities.ts
Comment (beets) [inline static/js/apps/base/utilities/utilities.ts:39]: curious when we encounter this string?

File: static/js/apps/base/components/HeaderBar.tsx
Comment (beets) [inline static/js/apps/base/components/HeaderBar.tsx:1]: please use snake case for these longer filenames e.g. header_bar.tsx

**Review by beets — COMMENTED** [review]
thanks for sending this change. in general, it's much easier for us if these pr's are small :)

the changes look reasonable and looks like it's at parity - even the e.g. ?hl=fr parameter works! i didn't review the homepage components too closely since we are about to change those to the new content.

for the footer, it looks like we are doing away with the full footer and only keeping the bottom 4 links. so hopefully a lot of it can be simplified as well.

while we are adding code, can we please add file/module comments, and comments for exported functions. tests would also be great for the exported functions.

File: static/js/apps/base/utilities/utilities.ts
Comment (beets) [reply static/js/apps/base/utilities/utilities.ts:48]: +1 -- i'd prefer to fail loudly in dev for errors like that (and in getLabels)

Comment (beets) [reply static/js/apps/base/utilities/utilities.ts]: I see - thanks for explaining the reasoning behind the logic. In that case, how about moving all the routing logic up into jinja and keeping the routes in the JSON config. The ts should only include fully formed URLs (with `{tools.visualization}#visType=timeline` expanded out as well), removing the need for regex's and onload complexity?

File: static/js/apps/homepage/components/learn_more.tsx
Comment (dwnoble) [inline static/js/apps/homepage/components/learn_more.tsx:48]: "{static.about}"

File: static/js/apps/homepage/components/tools.tsx
Comment (dwnoble) [inline static/js/apps/homepage/components/tools.tsx:73]: "{place.place}"

File: static/js/apps/homepage/components/tools.tsx
Comment (dwnoble) [inline static/js/apps/homepage/components/tools.tsx:81]: "{tools.download}"

File: static/js/apps/base/components/footer.tsx
Comment (dwnoble) [inline static/js/apps/base/components/footer.tsx:104]: Remove extra logo

File: static/js/apps/base/utilities/utilities.ts
Comment (beets) [reply static/js/apps/base/utilities/utilities.ts:53]: thanks - could you add a todo here to revisit this after the revamp?

File: static/js/apps/base/utilities/utilities.ts
Comment (beets) [reply static/js/apps/base/utilities/utilities.ts:29]: sounds good - could you add a todo here to revisit this after we get the bulk of the revamp done?

File: static/js/apps/homepage/app_v2.tsx
Comment (beets) [inline static/js/apps/homepage/app_v2.tsx:18]: please add a todo to fold this into app.tsx (dropping the v2) when ready.

**Review by beets — APPROVED** [review]
thanks for addressing the comments. i left a few comments - they are mainly to add todo's in code so it's easy to keep track of tasks we are deferring for now in the interest of time. we'll have to do another pr of all the changes when we merge to the main branch, so we'll take a closer look at everything then. i'll approve now to unblock progress - please merge after the todo's are added.

**Comment by dwnoble** [comment]
Thanks @nick-next and @jenibelle08!

**Pull request merged**
